### PR TITLE
Email outbox backend

### PR DESCRIPTION
**Pull request by N2D4**
I'm sorry for whoever reviews this. I tried to split it up as much as possible so this is only the existing functionality migrated onto the new email outbox table; the new endpoints or UI changes are not here yet.

Tests not passing yet but already ready for reviews.

How to review:

- The most important file is schema.prisma. It contains the explanation of the new EmailOutbox table, and all its constraints.
- After that, check the updated emails.tsx. It writes to the EmailOutbox table.
- Then, check email-queue-step.tsx. It contains the logic that renders, queues, and sends an email (alongside emails-low-level.tsx). This is called once every ~second (in a Vercel cron job in prod or a script in dev) and whenever a new email is sent (see the callers of runEmailQueueStep for more info).
- Then, look at the updated endpoints. Would love to have some eyes on the ones that @BilalG1 wrote in particular.
- Finally, everything else. Everything else should be relatively straight-forward, I think, mostly migrations.

The prompt that generated the first version (but note that the final version is somewhat different from what was generated, but it might help get some overview):

```
i want you to implement a new email sending backend infrastructure. apps/backend is the right folder

let's start with the database. For this purpose, I've already created an EmailOutbox model in the prisma schema. Note that, for now, canHaveDeliveryInfo is always false (we never fetch delivery info rn). Note that the EmailOutbox is on the global database, not the source-of-truth database (which means it's not in the same database as the project users, for example, and can't have foreign key relations to those). You must create a new migration for this new schema; you will have to manually write the migration SQL for it (can't use the Prisma command). After that, create another subsequent migration that migrates from the old SentEmail table to the new one.

Create a file in `lib` that has functions for the email outbox for listing all emails (returning the database objects), editing an email, resetting an email, deleting an email (which should only work under the same conditions as the retry one). This file should be SUPER clean, make sure to have many explained/documented functions with concise docs.

You also want to update emails.tsx to use the new email outbox. The usual current email-sending functions should instead of sending the email directly, create an EmailOutbox row in the database instead. It should also queue the email-queue-step below onto QStash so that it can start queuing itself repeatedly (this should probably be exported as a function so. Keep the existing function implementations somewhere for your own reference until you've implemented the email queue step endpoint down below, as you'll need to implement the email queue step endpoint down below with them. Note that if you send from a templtae to many users, each user is its own row in the email outbox system. Feel free to update the signature of the send email functions to match the information that is required in EmailOutbox, just update all the callers.

Also, update the `email-rendering.tsx` file to export a new function that allows you to render many email templates/drafts/themes at once, as long as they are for the same tenancy ID. They should all be in a single call to Freestyle.

Note that difference between InternalError and ExternalError in the EmailOutbox schema. Internal errors contain all the information, but maybe have sensitive implementaiton-speciifc data that should not exposed to the end user who sent the email. External errors are guaranteed to be safe to expose to the end user who sent the email. There is already a similar handling that's implicit in the emails.tsx file, so take some of its logic and make it more explicit (for both rendering & sending server errors, make functions that take internal full error objects and returns an external error object, with just a string like "Internal server error" if it's an unknown error). 

Also, create the following endpoints (please look at other endpoints in the backend to see how they're implemented, and make sure the security is good):

- GET /api/latest/emails/delivery-info: This endpoint returns some stats on the emails. It should return the number of emails sent in the last hour, last day, last week, and last month, and how many of those emails (per hour, day, week, and month) were marked as spam or bounced. It should also return the "capacity rate": the number of emails that can be sent per second. It should be calculated like this: `max(max(<the number of emails sent in the last week> * 4, 200) * penaltyFactor, 5) / 7 / 60 / 60`, where penaltyFactor is a number between 0 and 1 depending on what % of emails were bounced or marked as spam (I will build a precise metric for this later, for now you can think about it and make your own).
- GET /api/latest/emails/outbox: List all emails in the outbox, paginated like the users endpoint. Should return a nice type for the individual rows which should at least contain the status and simple status, and probably some other useful fields that you think should be displayed to the user in the UI, but not all those from the database (if you're not sure whether it's needed, don't include it, we can include it later). The function which converts DB EmailOutbox objects to the API type should be separate from the other stuff as we will use it elsewhere too.
- PATCH /api/latest/emails/outbox/{id}: Edit an email in the outbox, if the conditions for it are met. Only some fields can be edited, see the Prisma schema for an explanation
- POST /api/latest/emails/outbox/{id}/cancel: Cancel an email in the outbox, if the conditions for it are met
- POST /api/latest/emails/outbox/{id}/retry: Retry an email in the outbox, if the conditions for it are met. See the Prisma schema for an explanation
- POST /api/latest/internal/email-queue-step: This is a QStash receiver that is constantly called by QStash in an infinite loop. Make sure that it is super clean and easy to understand. Feel free to split it up into multiple functions and files as needed.
- - First thing it should do is add a call to itself back into the QStash queue, with Flow Control with a flow control key equal to the URL, a rate limit of 1 per second, and a parallelism limit of 1.
- - Then, it should acquire a Postgres exclusive lock, to make sure that only one email-queue-step can be running at a time (make sure this is easy to understand and free of data races). It should be released automatically at the end of this endpoint (think about how we can make it that there's no deadlock if the endpoint fails for whatever reason, eg. the server it's running on dies).
- - Now, it should get the timestamp of the last execution of this endpoint (which is stored in some metadata table in the global database), and calculate the precise time delta in seconds. The new time should be updated in the metadata table. (Make sure that it's not negative, which can happen due to clock skew, so make sure to handle that by just not doing anything if the new time is before the last execution time).
- - Then, it should fetch all the email outbox rows that do not have a renderedByWorkerId, and set renderedByWorkerId to a random UUID and startedRenderingAt to now (it should do this in an atomic statement so that an email is never rendered by multiple workers at the same time). It should then use email-rendering.tsx to render the email, using the previously created batch-by-tenancy function (you need to group the email outbox rows by tenancy ID for this). Once done, it should (again, in a single atomic statement) check whether the renderedByWorkerId is still set to the same UUID (this is to make sure there are no race conditions), and if so, set the appropriate fields in the EmailOutbox row.
- - Next, it should run a SQL query that updates all rows with scheduledAt < now() and isQueued = false to set isQueued = true.
- - Now, it should fetch a deduplicated list of all tenancy IDs of all email outbox rows that have isQueued = true and startedSendingAt being null, and set startedSendingAt to now.
- - For each tenancy ID in the list:
- - - Calculate the capacity rate. Multiply it by the time delta since the last endpoint execution. Use stochastic rounding to round to the nearest integer.
- - - For that tenancy, fetch at most that many email outbox rows that have isQueued = true and startedSendingAt being null, and set startedSendingAt to now. Fetch those rows that have the highest priority (see the Prisma schema for an explanation).
- - - Then, asynchronously in parallel, process each email that needs to be sent. Use the email-sending functions to send the emails, and update finishedSendingAt and canHaveDeliveryInfo (recall it's always set to false for now) for each row when the email is sent. Before you send the email, check whether the user still exists in the DB, and use the same notification category logic from the send-email endpoint (extract it into a helper function) to find out which users have already unsubscribed and should not receive the email, and set skippedReason accordingly if necessary. If the email encountered an error while sending, set the sendServerError-ish fields in the same atomic statement. `waitUntil` each promise (to make sure that they continue running in the background). Ensure that a bug when handling one of the emails does not mess with the other emails. Don't use `.catch`, ever, use try-catch blocks, and use `captureError("<some-machine-readable-location-id>", ...)` to capture all unknown errors aggressively.


While you're at it, we now have some new features like scheduling, add those to the send-email endpoint.

Make sure to make VERY comprehensive E2E tests for ALL the edge cases.

The code should be ULTRA clean and NEVER repetitive, feel free to create helper functions and files as needed. Definitely make sure everything is clean and easy to understand. When there's some logic that is used in multiple places, create a helper function for it. In the end, write a human-readable README.md file in the email endpoint folder that explains how the email infrastructure works, meant to be read by a technical human (it should be VERY concise and understandable!)

```

---

> [!NOTE]
> Moves email sending to a new EmailOutbox with async render/queue/send, adds delivery metrics, internal queue step endpoint/runner, and updates APIs/SDK/tests accordingly.
> 
> - **Database**:
>   - Add `EmailOutbox` schema with statuses, constraints, indices; introduce `EmailOutboxProcessingMetadata`.
>   - Migrate legacy `SentEmail` rows into `EmailOutbox` and drop `SentEmail`.
> - **Email pipeline**:
>   - Replace direct sends with outbox enqueue via `sendEmailToMany`; add recipient modeling and scheduling/priority.
>   - Implement queue worker `runEmailQueueStep` (render → queue → send) with capacity control and retries.
>   - Add low-level sending module (`emails-low-level`) and batched Freestyle rendering (`renderEmailsForTenancyBatched`).
> - **APIs**:
>   - New `GET /api/latest/emails/delivery-info` returning per-window stats and capacity.
>   - New `GET /api/latest/internal/email-queue-step` for cron-driven processing; add local runner script.
>   - Update send flows (OTP, password reset, invitations, verification) to use default template enqueue and allow deliverability bypass.
>   - Internal emails list now reads from `EmailOutbox`.
> - **SDK**:
>   - Expose `getEmailDeliveryStats()` and types; wire through server app caches.
> - **Tooling & ops**:
>   - Add Vercel cron for queue step; dev script to run queue; switch builds to ESM where needed; minor OpenAPI import fix.
>   - Docker: add Drizzle Gateway; tweak Postgres init.
> - **Tests**:
>   - Extensive E2E for delivery-info, queue step, API keys, email flows; adjust helpers for mailbox/waits.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e58e84296a69c9c482b314591ab1119145426e8c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Email delivery statistics endpoint showing sent, bounced, and marked-as-spam metrics across multiple time windows.
  * Background email queue processing system for asynchronous email delivery with rendering and sending orchestration.
  * High-priority email support for critical communications.
  * Email capacity tracking and rate limiting information.

* **Improvements**
  * Simplified email sending API response structure.
  * Enhanced deliverability check configuration options.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: apps/backend/prisma/schema.prisma [hunk 1]
```diff
@@ -661,28 +661,175 @@ model EventIpInfo {
...
+  simpleStatus EmailOutboxSimpleStatus @default(dbgenerated("<manually set>"))
+
+  // priority is the sending priority of the email among all emails that are not yet sent but already past their scheduled time. Higher priority means it should be sent sooner.
+  //
+  // This column is auto-generated based on the following formula:
+  //  priority = (isHighPriority ? 100 : 0) + (renderedIsTransactional ? 10 : 0)
+  priority Int @default(dbgenerated("<manually set>"))
+
+  // This column is auto-generated as defined in the SQL migration. It can not be set manually.
+  isPaused Boolean @default(false)
```
Comment (BilalG1) [inline apps/backend/prisma/schema.prisma]: I think we can remove the comment here since it isn't generated

File: apps/backend/src/lib/emails.tsx [hunk 2]
```diff
@@ -27,399 +37,63 @@ function getDefaultEmailTemplate(tenancy: Tenancy, type: keyof typeof DEFAULT_TE
...
-  }
-
-  await sendEmail({
-    tenancyId: options.tenancy.id,
-    emailConfig: await getEmailConfig(options.tenancy),
-    to: options.email,
-    subject: result.data.subject ?? "",
-    html: result.data.html,
-    text: result.data.text,
+  const themeSource = getEmailThemeForThemeId(options.tenancy, template.themeId);
```
[the remark this answers is not in the document]
Comment (BilalG1) [reply apps/backend/src/lib/emails.tsx]: good catch, themeSource unused

File: apps/backend/src/app/api/latest/integrations/credential-scanning/revoke/route.tsx [hunk 3]
```diff
@@ -184,18 +183,8 @@ export const POST = createSmartRouteHandler({
...
-    for (const email of affectedEmails) {
-      await sendEmail({
-        tenancyId: updatedApiKey.tenancyId,
-        emailConfig,
-        to: email,
-        subject,
-        html: htmlContent,
-      });
-    }
+    throw new StackAssertionError("Credential scanning email is currently disabled!");
```
Comment (BilalG1) [inline apps/backend/src/app/api/latest/integrations/credential-scanning/revoke/route.tsx]: Why can't we send email here?

File: apps/backend/src/app/api/latest/internal/email-queue-step/route.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  }),
+  response: yupObject({
+    statusCode: yupNumber().oneOf([200]).defined(),
+    bodyType: yupString().oneOf(["json"]).defined(),
+    body: yupObject({
+      ok: yupBoolean().defined(),
+    }).defined(),
+  }),
+  handler: async (_req, fullReq) => {
+    const startTime = performance.now();
```
Comment (BilalG1) [inline apps/backend/src/app/api/latest/internal/email-queue-step/route.tsx]: need to verify vercel cron secret here

File: apps/backend/src/lib/notification-categories.ts [hunk 5]
```diff
@@ -48,13 +52,12 @@ export const hasNotificationEnabled = async (tenancy: Tenancy, userId: string, n
...
-    method: {
-      email: "test@test.com",
-      type: "standard",
+    method: {},
+    data: {
+      user_id: userId,
+      notification_category_id: notificationCategoryId,
     },
     callbackUrl: undefined,
   });
```
[the remark this answers is not in the document]
Comment (BilalG1) [reply apps/backend/src/lib/notification-categories.ts]: should we keep expiresInMs here

File: apps/backend/src/app/api/latest/emails/send-email/route.tsx [hunk 6]
```diff
@@ -20,11 +16,14 @@ type UserResult = {
 const bodyBase = yupObject({
   user_ids: yupArray(yupString().defined()).optional(),
   all_users: yupBoolean().oneOf([true]).optional(),
-  subject: yupString().optional(),
-  notification_category_name: yupString().optional(),
+  subject: yupString().optional(),  // TODO unused rn; fix this
+  notification_category_name: yupString().optional(),  // TODO unused rn; fix this
```
[the remark this answers is not in the document]
Comment (BilalG1) [reply apps/backend/src/app/api/latest/emails/send-email/route.tsx]: yeah we should handle these. I think we can pass into `createTemplateComponentFromHtml` and then render our custom components

File: apps/backend/src/app/api/latest/internal/email-queue-step/route.tsx [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import { runEmailQueueStep } from "@/lib/email-queue-step";
+import { createSmartRouteHandler } from "@/route-handlers/smart-route-handler";
+import { yupBoolean, yupNumber, yupObject, yupString } from "@stackframe/stack-shared/dist/schema-fields";
+import { wait } from "@stackframe/stack-shared/dist/utils/promises";
+
+export const GET = createSmartRouteHandler({
```
[the remark this answers is not in the document]
Comment (BilalG1) [reply apps/backend/src/app/api/latest/internal/email-queue-step/route.tsx]: !

[the remark this answers is not in the document]
Comment (N2D4) [reply apps/backend/src/app/api/latest/internal/email-queue-step/route.tsx]: thx, i got rid of the email queue

[the remark this answers is not in the document]
Comment (N2D4) [reply apps/backend/src/app/api/latest/internal/email-queue-step/route.tsx]: (I mean I got rid of the one on Upstash, it's all with Cron jobs now)

File: apps/backend/src/app/api/latest/internal/email-queue-step/route.tsx (hunk 4 above)
Comment (N2D4) [reply apps/backend/src/app/api/latest/internal/email-queue-step/route.tsx]: good catch thx

File: apps/backend/src/app/api/latest/integrations/credential-scanning/revoke/route.tsx (hunk 3 above)
Comment (N2D4) [reply apps/backend/src/app/api/latest/integrations/credential-scanning/revoke/route.tsx]: We totally could, I just didn't implement it since this endpoint is currently unused (and has been for a while)

File: apps/backend/src/lib/email-queue-step.tsx [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+  const requests = await Promise.all(group.map(async (row) => {
+    const themeSource = getEmailThemeForThemeId(tenancy, row.themeId ?? false);
+
+    const recipient = deserializeRecipient(row.to as Json);
+    let userDisplayName: string | null = null;
+    let unsubscribeLink: string | undefined;
+    if ("userId" in recipient) {
+      const user = userMap.get(recipient.userId);
+      userDisplayName = user?.displayName ?? null;
+      if (row.renderedNotificationCategoryId) {
```
Comment (N2D4) [inline apps/backend/src/lib/email-queue-step.tsx]: TODO this is never true... need to fix by rendering twice

File: apps/backend/src/app/api/latest/internal/emails/crud.tsx [hunk 9]
```diff
@@ -33,7 +35,7 @@ export const internalEmailsCrudHandlers = createLazyProxy(() => createCrudHandle
   onList: async ({ auth }) => {
     const prisma = await getPrismaClientForTenancy(auth.tenancy);
 
-    const emails = await prisma.sentEmail.findMany({
+    const emails = await prisma.emailOutbox.findMany({
```
[the remark this answers is not in the document]
Comment (N2D4) [reply apps/backend/src/app/api/latest/internal/emails/crud.tsx]: gang observation
